src/TestFramework/Coverage/JUnit/JUnitTestCaseSorter.php

```diff
@@ -36,53 +36,134 @@
 namespace Infection\TestFramework\Coverage\JUnit;
 
 use function array_key_exists;
+use function count;
 use function current;
 use Infection\AbstractTestFramework\Coverage\TestLocation;
+use function Safe\ksort;
 use function Safe\usort;
 
 /**
  * @internal
  */
 final class JUnitTestCaseSorter
 {
+    /**
+     * Expected average number of buckets. Exposed for testing purposes.
+     *
+     * @var int[]
+     */
+    public const BUCKETS_COUNT = 25;
+
+    /**
+     * For 25 buckets QS becomes theoretically less efficient on average at and after 15 elements.
+     * Exposed for testing purposes.
+     */
+    public const USE_BUCKET_SORT_AFTER = 15;
+
     /**
      * @param TestLocation[] $tests
      *
-     * @return string[]
+     * @return iterable<string>
      */
     public function getUniqueSortedFileNames(array $tests): iterable
     {
         $uniqueTestLocations = $this->uniqueByTestFile($tests);
 
-        if (count($uniqueTestLocations) === 1) {
+        $numberOfTestLocation = count($uniqueTestLocations);
+
+        if ($numberOfTestLocation === 1) {
             // Around 5% speed up compared to when without this optimization.
             /** @var TestLocation $testLocation */
             $testLocation = current($uniqueTestLocations);
 
+            /** @var string $filePath */
             $filePath = $testLocation->getFilePath();
 
-            if ($filePath !== null) {
-                yield $filePath;
-            }
-
-            return;
+            return [$filePath];
         }
 
         /*
+         * We need to sort tests to run the fastest first.
+         *
          * Two tests per file are also very frequent. Yet it doesn't make sense
          * to sort them by hand: usort does that just as good.
          */
 
-        // sort tests to run the fastest first
-        usort(
-            $uniqueTestLocations,
-            static function (TestLocation $a, TestLocation $b) {
-                return $a->getExecutionTime() <=> $b->getExecutionTime();
-            }
+        if ($numberOfTestLocation < self::USE_BUCKET_SORT_AFTER) {
+            usort(
+                $uniqueTestLocations,
+                static function (TestLocation $a, TestLocation $b) {
+                    return $a->getExecutionTime() <=> $b->getExecutionTime();
+                }
+            );
+
+            return self::sortedLocationsGenerator($uniqueTestLocations);
+        }
+
+        /*
+         * For large number of tests use a more efficient algorithm.
+         */
+        return self::sortedLocationsGenerator(
+            self::bucketSort($uniqueTestLocations)
         );
+    }
+
+    /**
+     * Sorts tests to run the fastest first. Exposed for benchmarking purposes.
+     *
+     * @param TestLocation[] $uniqueTestLocations
+     *
+     * @return iterable<TestLocation>
+     */
+    public static function bucketSort(array $uniqueTestLocations): iterable
```

Comment:
If it needs to be exposed, we might as well extract it in a dedicated class (which could have a private constructor for)?

Reply:
Fair. How's TestLocationBucketSorter working for a name?

Reply:
👌